regex-capi/src/rure.rs

```diff
@@ -221,11 +226,77 @@ ffi_fn! {
             Err(_) => return -1,
             Ok(name) => name,
         };
+        println!("{}", name);
         re.capture_names.get(name).map(|&i|i).unwrap_or(-1)
     }
 }
 
 ffi_fn! {
+    fn rure_iter_capture_names_new(
+        re: *const Regex,
+    ) -> *mut IterCaptureNames {
+        let re = unsafe { &*re };
+        Box::into_raw(Box::new(IterCaptureNames {
+            re: re,
+            capture_names: re.re.capture_names(),
+        }))
+    }
+}
+
+ffi_fn! {
+    fn rure_iter_capture_names_free(it: *mut IterCaptureNames) {
+        unsafe { Box::from_raw(it); }
+    }
+}
+
+ffi_fn! {
+    fn rure_iter_capture_names_next(
+        it: *mut IterCaptureNames,
+        capture_name: *mut c_char,
+    ) -> bool {
+        println!("Entering rure_iter_capture_names_next");
+        // Assignments
+        let it = unsafe { &mut *it };
+        let re = unsafe { &*it.re };
+
+        // Safety checks
+        if capture_name.is_null() {
+            return false;
+        }
+
+        // Find next name
+        let cn = match it.capture_names.next() {
+            // Top-level iterator ran out of capture groups
+            None => return false,
+            Some(val) => {
+                let name = match val {
+                    // inner Option didn't have a name
+                    None => "",
+                    Some(name) => name
+                };
+                name
+            }
+        };
+        println!("Selected name: {}", cn);
+
+        // Fill in pointer
+        unsafe {
+            let cs = match CString::new(cn.as_bytes()) {
+                Result::Ok(val) => {
+                    val
+                },
+                Result::Err(err) => return false
+            };
+            println!("CString cast: '{}'", cs.to_str().unwrap());
+            *capture_name = *cs.into_raw();
```

Comment:
I think the problem is here; the `CString` is getting freed when the `unsafe` block completes, thus the C unit test triggers a segfault trying to read the memory (similar to https://users.rust-lang.org/t/correct-way-to-implement-a-function-which-returns-a-c-string/315/3 ). I think the solution is to use a `Box`, but I'm unsure of the details of how to do that.